Create Annotator class

kerimoyle opened the pull request:
This is a work in progress to create tools that would support developers of software to annotate CellML models #614 

- [x] retrieve item by id from model (type = std::any),
- [x] retrieve item-of-type from model,
- [x] retrieve item type by id from model,
- [x] set ids on everything (addressed separately in #639 ).
- ~[ ] fix ambiguity around returning connection items (should be specified by component pair not variable pair?)~

kerimoyle committed:
Blank Annotator class, building

kerimoyle committed:
WIP: returns using variant but can't retrieve active value

kerimoyle committed:
Tests passing

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle commented:
Failing on the Windows build? What do I do about it @hsorby?

agarny commented:
> Failing on the Windows build? What do I do about it @hsorby?

It looks like we are using MSVC 2015 and `<any>` is only available in MSVC 2017 (see [here](https://devblogs.microsoft.com/cppblog/c17-features-in-vs-2017-3/)).

kerimoyle committed:
Changes as discussed

- return AnyItem from item(id) function, remove overloads
- use enum for types of return items
- error handling/casting tests
- connection tests include any variable pairs

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
experimenting: returning instructions or index strings to items

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -182,3 +182,446 @@ TEST(Annotator, errorHandling)
+        "map_variables_2",
+        "connection_1",
+    };
+    std::vector<std::string> expectedInstructions = {
```
Hey @nickerso Would these work for your instructions?

kerimoyle replied:
Removing after discussion last week.

kerimoyle commented:
Work in progress exploring different ways for users to interact with the `Annotator` class:
- retrieving an enum type and `std::any` item; user must cast item manually;
- retrieve an enum type and set of indices; user can retrieve from the index chain; 
```
Annotator::Type::COMPONENT, {1, 2, 3}  <-->  model->component(1)->component(2)->component(3)
Annotator::Type::VARIABLE, {1, 2, 3} <--> model->component(1)->component(2)->variable(3)
... etc ...
```
- retrieve a string containing instructions for how to reach the item.

kerimoyle committed:
Trigger new build

kerimoyle committed:
Made casting errors platform independent

kerimoyle committed:
reinstated options for item retrieval, coverage tests

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
reinstated docstrings for Annotator

kerimoyle committed:
renamed convenience functions to avoid import() as a name

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Removed functions as discussed

kerimoyle commented:
Ready for review please @agarny @hsorby @nickerso All changes made as discussed yesterday.

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

nickerso commented on src/annotator.cpp:
```diff
@@ -0,0 +1,361 @@
+{
+    std::string id = component->id();
+    if (!id.empty()) {
+        idList.insert(std::make_pair(id, std::make_pair(0, component)));
```
probably missing something, but shouldn't that `0` be `Annotator::Type::COMPONENT` ?

kerimoyle replied:
Yes, weirdness.  Thanks!

kerimoyle replied:
83ce6a11

nickerso commented on src/annotator.cpp:
```diff
@@ -0,0 +1,361 @@
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+    if (mPimpl->mIdList.empty()) {
```
Not sure this check is needed. If you can't find the requested `id`, it doesn't really matter whether there are no IDs or not. At the very least, I don't see how this is an error, if anything a hint or maybe a warning...but would just remove this block.

kerimoyle replied:
It needs to be there to distinguish between situations where the annotator has been built (the previous check) and where zero items with ids exist in the model.  Otherwise, we can't tell the difference between an empty idList.  It doesn't need to be an error, but it does need to be a return path.

kerimoyle replied:
83ce6a11

nickerso replied:
but why do we need to distinguish the case when there are no items with `id`s the model vs the requested `id` is not used? There is absolutely nothing wrong with that, so I'm not sure why we need to care?

This use of `any` is really looking less and less attractive, as it seems to be forcing us to do things that just don't make sense...

kerimoyle replied:
There's not a lot of difference, other than telling the user that there's no point looking for any ids, as there are none, rather than just not having the single one it's looking for.  I can lump them together if it's clearer.

kerimoyle replied:
83ce6a11

nickerso commented on src/annotator.cpp:
```diff
@@ -0,0 +1,361 @@
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        // Adding the requested id string and this issue to the map, so that future requests (ie: to the .second)
+        // will return the same IssuePtr rather than creating a new one.
+        mPimpl->mIdList.insert(std::make_pair(id, item));
```
no, really don't like this approach. Essentially adding a new ID to the model in a not very nice way. But also looking at this, it means that this method will always return an `item` and then the user needs to check if that item is of type `ISSUE` to see if it was successful. That seems very wrong to me...but maybe I'm missing something about those any items...

kerimoyle replied:
No, that's not how it works - there's no ID added to the model at all.  This is used so that we can keep track of the same request for nonexistent IDs, and only return one issue for them.  Otherwise we'd be generating a duplicate issue each time.

nickerso replied:
yeah, sorry, not really added to the model, but if you just look at the id list in the annotator as the list of items in the current model that can be annotated the user is forced to check that each item is at least not an `issue` before being able to use it...again, this is starting to look like the use of `any` is less than optimal.

kerimoyle replied:
It's nothing to do with the use of `any`? That's only so that we can return a pointer reasonably efficiently (rather than having a list of each different pointer type in the argument list of the function).  It's easy enough to remove!  We could just as easily have an `Annotator::Type::NOT_FOUND` enum or similar, with a `nullptr` as the object.  It would still need the user to check for type though, same as before.  But a side note - is viewing/printing a list of ids something a user would want?  I don't do that at the moment?

nickerso commented on src/annotator.cpp:
```diff
@@ -0,0 +1,361 @@
+        mPimpl->mIdList.insert(std::make_pair(id, item));
+        return item;
+    }
+    return mPimpl->mIdList[id];
```
In most (all?) methods we have where the user requests an object by "name", you get a `nullptr` if that name doesn't exist or an object if it does. Can we not do that here?

kerimoyle replied:
Alas, no, for two reasons.  I tried this originally, and I couldn't cast a nullptr into the required return type that the compiler would accept.  Given that, it seemed better to return something that would be helpful to the user, and prevent duplicate issues being generated (as above).

kerimoyle replied:
If they request a specific item type they will get a `nullptr` (eg: annotator->component("iDontExist") will return `nullptr`.  It's only the `item` function that is different.

nickerso commented on src/annotator.cpp:
```diff
@@ -0,0 +1,361 @@
+    return mPimpl->mIdList[id];
+}
+
+ComponentPtr Annotator::component(const std::string &id)
```
I can see these methods all look like they should work, but would be keen to know if the try/catch approach is what we should be using rather that testing the item type. @hsorby @agarny

nickerso replied:
could be other reasons for the cast to fail beyond just the item type. So this seems like a reasonable way to go and will ensure that the casting exceptions would ever propagate out from the library

nickerso commented on src/api/libcellml/annotator.h:
```diff
@@ -0,0 +1,289 @@
+     *
+     * @param model A @c ModelPtr model to build map for.
+     */
+    void build(const ModelPtr &model);
```
can we leverage the `build` method to add IDs to any objects that don't have them? Seems like that capability should be somewhere in the annotator class rather than (as well as?) the current addition of IDs in the printer.

nickerso commented on src/api/libcellml/annotator.h:
```diff
@@ -0,0 +1,289 @@
+     *  - an @c Annotator::Type enum, and
+     *  - an @c std::any item containing the item.
+     *
+     * Possible string labels and their corresponding items returned are:
```
type enums rather than string labels, right?

nickerso commented on src/api/libcellml/annotator.h:
```diff
@@ -0,0 +1,289 @@
+          *
+     * @param id A @c std::string representing the @p id to retrieve.
+     *
+     * @return An @c AnyItem item as described above.
```
Not at all endorsing the "return an issue type AnyItem on error" - but if that remains it should be clear in the docs here that is what would happen.

nickerso commented on src/utilities.cpp:
```diff
@@ -556,4 +559,122 @@ void findAllVariablesWithEquivalences(const ComponentPtr &component, VariablePtr
     }
 }
 
+IdList listIds(const ModelPtr &model, bool mathIds)
```
seems to be very similar code here and in the annotator...any way to merge them?

nickerso commented on src/utilities.h:
```diff
@@ -211,6 +211,22 @@ static const std::map<Variable::InterfaceType, const std::string> interfaceTypeT
     {Variable::InterfaceType::PUBLIC, "public"},
     {Variable::InterfaceType::PUBLIC_AND_PRIVATE, "public_and_private"}};
 
+// static const std::map<Annotator::Type, const std::string> annotatorTypeToString = {
```
Is this commented code still needed?

kerimoyle replied:
Nope :)

kerimoyle replied:
83ce6a1

nickerso commented on src/utilities.h:
```diff
@@ -555,4 +571,9 @@ static inline std::string trimCopy(std::string s)
     return s;
 }
 
+// KRM need docstrings.
```
yes.  :)

kerimoyle replied:
These are slightly different in the PR which uses them more #639 (we've now dropped testing/adding mathml ids, for example).  Should I merge that one with this one first?  It seems odd to change it in two places (which will have knock-on effects throughout the printer too).

kerimoyle replied:
Will be done in the other PR instead.

nickerso commented on tests/model/model.cpp:
```diff
@@ -784,11 +784,11 @@ TEST(Model, removeComponentInsensitiveToOrder)
     // If the order of operations is switched the behaviour is the same:
 
-    // Get a pointer to the second componet in the parsed model
+    // Get a pointer to the second componet in the parsed model.
```
`componet` ==> `component` ?

kerimoyle replied:
Serves me right for adding the missing full-stop ... ;)

kerimoyle replied:
83ce6a11

nickerso commented on tests/printer/printer.cpp:
```diff
@@ -295,3 +293,257 @@ TEST(Printer, printModelWithStandardUnitsAdded)
+                           "    </component_ref>\n"
+                           "  </encapsulation>\n"
+                           "</model>\n";
+    const std::string expectedOutString = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
```
what happened to the random/hexadecimal IDs?

kerimoyle replied:
They are in a different PR, see https://github.com/cellml/libcellml/pull/639

nickerso commented on src/utilities.cpp:
```diff
@@ -556,4 +559,122 @@ void findAllVariablesWithEquivalences(const ComponentPtr &component, VariablePtr
+
+std::string makeUniqueId(const std::string &type, IdList &idList)
+{
+    // Construct a unique id appropriate to the type, and add it to the idList.
```
I thought we went away from these type-based IDs and we just using hexadecimal strings?

kerimoyle replied:
We are, but that is a different PR.  See https://github.com/cellml/libcellml/pull/639

kerimoyle committed:
Some changes made as per review

nickerso reviewed (commented):
tentative approval :)

I'm not happy with the way the `item()` method is working, but it looks like that might just be something we need to live with for "version 1" of this functionality. Really seems that trying to make use of those `any` objects are forcing us down a path that is less that ideal...but it is better than what we have and provides useful functionality.

Happy to approve this subject to those C++ question(s) that I've put around the place...

nickerso commented:
(resolved some of my comments just cause they are not essential to address now, can be refined as we gain experience in using the Annotator)

kerimoyle committed:
Make empty list error message the same as unfound item

kerimoyle committed:
Update tests

kerimoyle committed:
Python bindings for Annotator, python tests

agarny commented on src/bindings/interface/annotator.i:
```diff
@@ -6,21 +6,19 @@
+
 %feature("docstring") libcellml::Annotator
-"Validates CellML objects.";
+"Interacs with CellML objects using their id attribute.";
```
Typo.

kerimoyle replied:
95ddcc31

kerimoyle committed:
Typo in docstring.

kerimoyle committed:
improve enums in python

kerimoyle committed:
Update annotator.cpp

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge conflicts fixed

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
WIP setting automatic ids through annotator

kerimoyle committed:
semicolons in interface files

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
WIP auto ids for individual items

needs coverage and custom interface/docstrings

kerimoyle committed:
Merge branch 'develop' into annotator_auto_ids

kerimoyle committed:
WIP bindings for python done, coverage not complete yet

kerimoyle committed:
Merge branch 'develop' into annotator_auto_ids

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
changed "set" to "assign"

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
removed repair function, added clearAll, assigning ids overwrites them

kerimoyle committed:
Merge branch 'annotator_auto_ids' of https://github.com/kerimoyle/libcellml into annotator_auto_ids

kerimoyle committed:
returning duplicates and tests, WIP

kerimoyle committed:
Merge branch 'develop' into annotator_auto_ids

kerimoyle committed:
WIP coverage and docstrings

kerimoyle committed:
returns automatic ids, docstrings, coverage

kerimoyle committed:
Changing binding for Ubuntu

kerimoyle committed:
Update annotator.i

kerimoyle committed:
Added overloads for bindings and tests

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge branch 'annotator_use_any_not_variant' into annotator_auto_ids

kerimoyle committed:
WIP hashing updates

kerimoyle committed:
basic hash working, tests

kerimoyle committed:
Updating hash for internal automatic ids

kerimoyle committed:
Allow broken ids to be assigned automatic ones

kerimoyle committed:
docstrings

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge branch 'develop' into annotator_auto_ids

kerimoyle committed:
Update interface file, tests pass

kerimoyle committed:
Fixed types.i

kerimoyle committed:
consolidate issues, build() renamed to buildModelIndex()

kerimoyle committed:
Merge branch 'annotator_auto_ids' into annotator_use_any_not_variant

kerimoyle committed:
make issue functions const

kerimoyle committed:
rearrange functions

kerimoyle committed:
renamed swig functions, formatting

kerimoyle committed:
cleanup

kerimoyle committed:
tidy ups

kerimoyle commented:
Hi @agarny @nickerso @hsorby This is ready for review now.  A couple of things:
- There is (necessarily) a lot of code duplication in the overloaded functions.  This is so that the users can call one function with anything and have it work as they'd expect - we have to do all the monkey business behind the scenes to make that possible.  The typecast restrictions mean that these have to be separate functions; I can't see any way to reasonably condense them.
- There's an awkward situation with the binding interface stuff because SWIG doesn't currently support the `std::any` type, but Python doesn't care ...  Thus, the `annotator.i` file defines some C++ functions with a `ForSWIG` extension in their names, which are there to provide an alternative to the real C++ functions in a format that can be digested by SWIG (and only by SWIG).  If there's a way to make these internal/private please let me know. At present they've got the same status as any other bound function.

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle commented:
ping @agarny @hsorby @nickerso

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

hsorby commented on tests/annotator/tests.cmake:
```diff
@@ -0,0 +1,16 @@
+#    ctest -R <category-label>
+# which will run the tests matching this category-label.
+# Can be left empty (or just not set)
+set(${CURRENT_TEST}_CATEGORY io)
```
I don't think the annotator belongs in the `io` category, maybe `tools`.

agarny replied:
We don't seem to have anything in the `tools` category, but we have some tests in the `utils` category, which I think would be fine for the annotator. Done in commit 5adb854b.

hsorby committed:
Refactor the annotator with use any.  Shift to using a weak pointer for the model.

hsorby reviewed (commented):
I initially intended to just do some indentation fixes and it just got out of control from there.

kerimoyle commented on src/api/libcellml/enums.h:
```diff
@@ -0,0 +1,46 @@
+    UNIT,
+    UNITS,
+    VARIABLE,
+};
```
Will we add `XML` and `INVALID` to this list so it can be used elsewhere?

hsorby replied:
Well, no I hope not.  The XML one is a problem but it is not a CellML element.  The `INVALID` should be covered by `UNDEFINED` no?

kerimoyle replied:
Ah, I thought yesterday you wanted an invalid type.  I'm fine with undefined to cover both.

kerimoyle replied:
We still need to rename this so that it includes `XML` and is named `ItemType` so that it can fit with the rest of the code... That will have knock-on effects throughout the rest of this PR.

nickerso replied:
is this part of #696? so we can resolve things on this PR and leave it to that one to sort out?

kerimoyle commented on src/api/libcellml/types.h:
```diff
@@ -16,20 +16,24 @@ limitations under the License.
 class Annotator; /**< Forward declaration of Annotator class. */
-using AnnotatorPtr = std::shared_ptr<Annotator>; /**< Type definition for shared annotator pointer. */
+using AnnotatorPtr = std::shared_ptr<Annotator>; /**< Type definition for @c std::shared Annotator pointer some othe text. */
 class Generator; /**< Forward declaration of Generator class. */
```
typo "othe" ... and also ... huh?

hsorby replied:
Playing around with Doxygen documentation.

kerimoyle commented on src/api/libcellml/types.h:
```diff
@@ -16,20 +16,24 @@ limitations under the License.
 class GeneratorVariable; /**< Forward declaration of GeneratorVariable class. */
 using GeneratorVariablePtr = std::shared_ptr<GeneratorVariable>; /**< Type definition for shared generator variable pointer. */
-class Importer; /**< Forward declaration of Importer class. */
+class Importer; /*!< Forward declaration of Importer class. */
```
Why the ! ?

hsorby replied:
Again playing with Doxygen documentation.

kerimoyle replied:
Are we leaving this as is ...? or changing it back?

kerimoyle commented on src/api/libcellml/types.h:
```diff
@@ -44,7 +48,7 @@ using ValidatorPtr = std::shared_ptr<Validator>; /**< Type definition for shared
 // CellML entities.
 class Component; /**< Forward declaration of Component class. */
-using ComponentPtr = std::shared_ptr<Component>; /**< Type definition for shared component pointer. */
+using ComponentPtr = std::shared_ptr<Component>; /**< Type definition for shared component pointer. @details Some extra definition for detailed description. */
```
... is this you just playing with doxygen translation stuff ... ?

hsorby replied:
Yes.  It should all be removed.

kerimoyle commented on src/bindings/interface/annotator.i:
```diff
@@ -104,13 +105,10 @@
-%feature("docstring") libcellml::Annotator::isBuilt
+%feature("docstring") libcellml::Annotator::hasModel
 "Report whether the annotator index is up-to-date with its stored model state.";
 
```
Docstring should be changed to reflect what the function does?

hsorby replied:
Definitely.

agarny replied:
This has been resolved.

kerimoyle commented on src/bindings/python/__init__.py:
```diff
@@ -51,22 +51,27 @@ class Object:
         'TEST_VALUE',
+        'UNDEFINED',
         'UNIT',
         'UNITS',
```
Also `XML` and `INVALID` ?

kerimoyle commented on src/bindings/python/__init__.py:
```diff
@@ -51,22 +51,27 @@ class Object:
+
+# The Annotator class relies on the CellMLElmement declaration above.
+from libcellml.annotator import Annotator
+
```
I don't understand why this is here?

hsorby replied:
Because of the transformation of how SWIG does enums and how we want to make them more friendly in Python.

kerimoyle replied:
... but does it need to be in this file?  Now that the enum for Type has been moved out of the Annotator class?

agarny replied:
What is the status of this file? I was going to remove the commented-out code, but then came across this unresolved conversation.

kerimoyle commented on src/internaltypes.h:
```diff
@@ -50,9 +49,17 @@ using VariablePtrs = std::vector<VariablePtr>; /**< Type definition for list of
+using ComponentWeakPtr = std::weak_ptr<Component>; /**< Type definition for weak component pointer. */
+//using ComponentEntityWeakPtr = std::weak_ptr<ComponentEntity>; /**< Type definition for weak component entity pointer. */
+//using EntityWeakPtr = std::weak_ptr<Entity>; /**< Type definition for weak entity pointer. */
+//using ImportedEntityWeakPtr = std::weak_ptr<ImportedEntity>; /**< Type definition for weak imported entity pointer. */
```
delete if not used?

hsorby replied:
Yes.

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -427,40 +445,24 @@ TEST(Annotator, errorHandling)
-    EXPECT_EQ(libcellml::Annotator::Type::ISSUE, annotator->item("i_dont_exist").first);
-    EXPECT_EQ(size_t(2), annotator->issueCount());
-    EXPECT_EQ(annotator->issue(0)->description(), "No model is stored; no action has been taken.");
-    EXPECT_EQ(annotator->issue(1)->description(), "The model index is out-of-date.  Please (re)build it before attempting to access items by their id.");
```
I don't understand why you're not testing which error is returned but only the total?

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -562,15 +564,19 @@ TEST(Annotator, getItemBySpecificTypeDuplicateId)
-    EXPECT_EQ(nullptr, annotator->encapsulation("duplicateId"));
+    EXPECT_NE(nullptr, annotator->model("duplicateId"));
+    EXPECT_NE(nullptr, annotator->encapsulation("duplicateId"));
+
```
I reeeaaaaeally don't agree with this change in behaviour!  I think it's really dodgy to not alert the user that the item found has a non unique id, and just return the first one we happen to come to.  This is not great ... :(

hsorby replied:
If the documentation of the method explains the behaviour then I think this is the correct.  The only thing I could see it doing is raising a warning that this identifier is not unique.

kerimoyle replied:
I think this will end up causing all kinds of problems.  It's naive and assumes that the users know our internal index order at which their components/variables/everything is stored ... which is kinda ridiculous as we ourselves don't ever rely on it!  Being able to interact with duplicated ids is a convenience function so that users can fix their own mistakes - I think it's dangerous to just hide the fact that it's duplicated from the user.  If they want to retrieve duplicated things they should use explicit functions to do that so that everyone knows exactly what they're dealing with.

nickerso replied:
is this one resolved? the methods seem to be documented to behave this way, so I'm assuming this is all good and we could come back and refactor this behaviour if needed...

hsorby replied:
I think so it is now back to the original setting, @kerimoyle can correct me if i'm wrong here.

kerimoyle replied:
Yes, I'm happier now, thanks.

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -1446,6 +1468,27 @@ TEST(Annotator, automaticIdAllItemsWrongModel)
+    model.reset();
+    EXPECT_FALSE(annotator->hasModel());
+
+    EXPECT_EQ("b4da55", annotator->assignId(itemComponent));
```
What happens when this component is added to a model? or another component without a parent model is passed to the annotator?  I think if the item passed in for id isn't a part of the stored model (or there is no stored model!) it should return an error instead of assigning an id which can't be said to be unique as it has no context.

hsorby replied:
Well if the component has no model then it is unique for the given context.  If you go and add it to a model the context has changed, nothing the library can do about that.

I don't see why it should return an error.  Someone has given me a component and they want an id assigned to it.  I can do that so I do, no problem.

kerimoyle replied:
The only reason to assign an id using the Annotator is to assign something unique.  As soon as there's no model, it can't make it unique.  So it's kinda pointless.

kerimoyle replied:
I will change the documentation so that it's clear that the model must be set first.

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -1513,10 +1556,10 @@ TEST(Annotator, automaticIdAllItemsEntityType)
         model->component("component2")->variable("variable2"));
 
-    EXPECT_EQ("b4da62", annotator->assignId(libcellml::Annotator::Type::CONNECTION, connection2));
+    EXPECT_EQ("b4da62", annotator->assignId(connection2));
```
I don't think we should have a default for variable pair stuff (connections and mappings).  And if we did, it shouldn't be the connection as we never actually interact with a connection directly.  But my strong preference would be there is no default at all.

hsorby replied:
Why would you not have a default?

kerimoyle replied:
A default is useful for real CellML entities, where the parent item (reset) is the default, and owns something else (the resetVale).  Here, there is no CellML entity, there is no ownership, and the thing which people interact with (the explicit mapping of **variables** to each other, not the implicit connecting of **connections** to each other) isn't the default.  If you insist on having one of them, as a default, it should be the mapping, not the connection.  But even that is a stretch for me.

kerimoyle replied:
Will change so that we have a default but to the `map_variables` instead of the connection.

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -1534,30 +1577,60 @@ TEST(Annotator, assignAllIds)
-    EXPECT_EQ(libcellml::Annotator::Type::ISSUE, retrievalError.first);
-    EXPECT_EQ(expectedErrors[1], std::any_cast<libcellml::IssuePtr>(retrievalError.second)->description());
+    libcellml::AnyItem item = annotator->item("duplicateId");
+    EXPECT_EQ(libcellml::CellMLElement::UNDEFINED, item.first);
```
Should also test that error is logged

agarny replied:
What do/did you mean by this @kerimoyle?

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -1911,42 +1984,21 @@ TEST(Annotator, hashChangesAndUpdates)
 
-    model->setId("iHaveChangedSinceTheLastBuild");
-    EXPECT_FALSE(annotator->isBuilt());
+    // Model is set and the id map is not created.
```
The id map is **not** created?? I thought it was built when the model was set??

agarny replied:
Is this comment still relevant? @kerimoyle? @hsorby?

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -1911,42 +1984,21 @@ TEST(Annotator, hashChangesAndUpdates)
 
-    // Shouldn't be able to retrieve anything since the build is out of date.
+    // The id map is built when a query is made.
     EXPECT_EQ(nullptr, annotator->model("model_1"));
```
This test is a bit odd now?  The id has changed, so you should rather test that the new id can now be found (ie: that the index has successfully detected the change and been rebuilt)?

agarny replied:
Is this comment still relevant? @kerimoyle? @hsorby?

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -1911,42 +1984,21 @@ TEST(Annotator, hashChangesAndUpdates)
+    // Can set id automatically, this does not trigger a rebuild of the id map.
+    EXPECT_EQ("b4da55", annotator->assignId(model->component(0)));
+
+    EXPECT_NE(nullptr, annotator->encapsulation("encapsulation_1"));
```
I don't understand this?  The old id should still be present and able to be retrieved since it's only the model id which was changed outside of the annotator?

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -1957,11 +2009,11 @@ TEST(Annotator, hashUpdatedWithAllAutomaticIds)
 
     annotator->assignAllIds();
-    EXPECT_TRUE(annotator->isBuilt());
+    EXPECT_TRUE(annotator->hasModel());
```
I understand the function names when `hasModel` is testing whether or not you have a model.  But I don't like it in this situation where it's actually testing that the stored model is up to date with the stored index.  Can we name it something else?

agarny replied:
Is this comment still relevant? @kerimoyle? @hsorby?

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -1970,78 +2022,64 @@ TEST(Annotator, autoIdOnOutOfDateBuild)
     libcellml::Variable::setEquivalenceConnectionId(connection.first, connection.second, "changed");
-    EXPECT_FALSE(annotator->isBuilt());
-    EXPECT_EQ("b4da57", annotator->assignId(libcellml::Annotator::Type::CONNECTION, connection));
+    EXPECT_EQ("b4da57", annotator->assignId(connection));
```
I think there shouldn't be a default for variable pair items (or if there is, it should be the map_variables one).  But better without any.

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -1970,78 +2022,64 @@ TEST(Annotator, autoIdOnOutOfDateBuild)
     model->component("component2")->reset(0)->setId("changed");
-    EXPECT_FALSE(annotator->isBuilt());
-    EXPECT_EQ("b4da5c", annotator->assignId(libcellml::Annotator::Type::RESET, model->component("component2")->reset(0)));
+    EXPECT_EQ("b4da5c", annotator->assignId(model->component("component2")->reset(0), libcellml::CellMLElement::RESET));
```
This should/could be with the RESET as default?

hsorby replied:
This is with RESET as default I am testing here that you can pass RESET in with out any problems.

nickerso commented on src/annotator.cpp:
```diff
@@ -0,0 +1,1741 @@
+    mPimpl->buildIdList();
+    mPimpl->mHash = mPimpl->generateHash();
+
+    for (auto it = mPimpl->mIdList.begin(), end = mPimpl->mIdList.end(); it != end; it = mPimpl->mIdList.upper_bound(it->first)) {
```
Why are duplicate ids being checked here? seems like a lot of work for something that the user hasn't asked for.

If it does stay here, then the documentation definitely needs to mention that this is happening that then the user should be checking for issues after calling this method.

kerimoyle replied:
That's a hangover from previous stuff I think, so it could go ... I'm still not very happy with how we're dealing with duplicates though.

kerimoyle replied:
57725d21

nickerso commented on src/annotator.cpp:
```diff
@@ -0,0 +1,1741 @@
+
+void Annotator::clearAllIds(ModelPtr &model)
+{
+    mPimpl->mModel = model;
```
won't this break everything? if anything, it should be calling `setModel` with the given `model`, right?

kerimoyle replied:
It doesn't make sense to call `setModel` which will build the expensive index, and then call `clearAllIds` to wipe it out.  I thought this was cleaner.

nickerso commented on src/api/libcellml/annotator.h:
```diff
@@ -0,0 +1,1054 @@
+    /**
+     * @brief Assign an id string to every item in the model which is already 
+     * stored in this annotator, excluding its MathML items.
+     * 
```
I thought previously this method would be adding ids to items that didn't already have one. If the user wants to assign to everything they would first clear and then assign. Now there is no method to easily make sure that everything has an id without losing existing ones.

kerimoyle replied:
Gah!! You told me to take that away!! I used to only assign a new id to things whose id was missing or invalid.  But, from our discussion, you said anyone wanting to assign ids automagically was prepared to nuke it all ... :(

nickerso replied:
hmmm...that doesn't sound like something I would say :) is there still a print method that will add ids to things? maybe that is the one that will do what I am thinking here...

but at the moment I'm struggling to think why it would ever be ok to ignore ids that have already been set.... I can see a case for wanting to specifically nuke them all and start from scratch, but not that it would be the only method to make sure everything has an id. It would be a fairly typical use-case to want to simply make sure that an existing model has ids everywhere so that, for example, a user can annotate the model without having to edit the model.

kerimoyle replied:
Haha, It's easy enough to change that back, but I'll wait until Hugh has finished his stuff.

kerimoyle replied:
8d4307d2

kerimoyle committed:
Remove check for duplicate ids at time of building

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
fixed initialisation, removed unique checks, updated test/format

kerimoyle committed:
Merge branch 'annotator_use_any_not_variant' of https://github.com/kerimoyle/libcellml into annotator_use_any_not_variant

kerimoyle committed:
automatic ids do not replace existing ids

kerimoyle committed:
Made casting test into a proper test

kerimoyle committed:
test to show issue #707

kerimoyle committed:
fix for issue #707

kerimoyle committed:
added ids() function to return all ids in a model, including test

kerimoyle committed:
docstring for bindings

kerimoyle committed:
Added dictonary functionality, and test

kerimoyle committed:
made typeAsString available in API

kerimoyle committed:
fixed iterator in duplicateIds

kerimoyle commented on src/annotator.cpp:
```diff
@@ -252,108 +435,80 @@ void Annotator::buildModelIndex(const ModelPtr &model)
-        return false;
-    }
-    return true;
+    return mPimpl->mIdList.count(id) == 1;
```
This needs to go back to the `> 1` test rather than `== 1`.  Otherwise we get a return of `false` when the id doesn't exist, which doesn't make sense...

kerimoyle committed:
Bindings overloaded

kerimoyle committed:
Fixed python tests since API changed

hsorby committed:
Make const strings const.

hsorby committed:
Merge branch 'annotator_use_any_not_variant' of github.com:kerimoyle/libcellml into kerimoyle/annotator_use_any_not_variant

hsorby committed:
Remove changes that were in place to test Doxygen output.

hsorby committed:
Improve documentation of unitId method.

hsorby committed:
Correct Python documentation for Annotator class.

hsorby committed:
Make additional Annotator class methods more Python private.

hsorby committed:
Refactor annotator class implementation.  Use file based models for Annotator tests.

hsorby committed:
Add some protections around the get id from named entities.

hsorby committed:
Add documentation for cellMLElementAsString.

hsorby committed:
Merging in changes from 'prime/develop' fixing conflicts.

hsorby committed:
Make changes required by clang format.

hsorby committed:
Save changes before commiting required by clang format.

hsorby committed:
Fix more clang format issues.

hsorby committed:
Apply clang format to codebase.

hsorby committed:
Improve detection of requirements for Python coverage.

hsorby committed:
Remove unnecessary check from buildMapsForComponentsVariables.

hsorby committed:
Improve the setting and retrieval of connection ids.

hsorby committed:
Merge branch 'annotator_use_any_not_variant' of github.com:kerimoyle/libcellml into kerimoyle/annotator_use_any_not_variant

hsorby committed:
fix: do not use 'else' after 'break'.

hsorby committed:
Fix: use range-based for loop.

hsorby committed:
Work on getting annotator Python coverage to 100%.  Includes working through some SWIG issues with std::pair and std::smart_ptr.

hsorby committed:
Peak of debugging for std containers and enums with SWIG.

hsorby committed:
Remove dictionary method from Annotator class.

hsorby committed:
Merge in changes from 'prime/develop' fixing conflicts.

hsorby committed:
Fix code formatting issues.

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

agarny commented on cmake/environmentchecks.cmake:
```diff
@@ -33,19 +33,30 @@ find_program(LLVM_PROFDATA_EXE NAMES ${PREFERRED_LLVM_PROFDATA_NAMES} llvm-profd
+      message(STATUS "Performing Test HAVE_COVERAGE - ${_RESULT_TEXT}")
     endif()
+    set(PYTHON_COVERAGE_FOUND ${HAVE_COVERAGE} CACHE INTERNAL "Python coverage was found." FORCE)
+    set(PYTHON_COVERAGE_TEST_PERFORMED TRUE CACHE INTERNAL "Testing for Python coverage has occured.")
```
Why do we have all of this which has nothing to do with creating an `Annotator` class?

hsorby replied:
That is indeed true, it also shouldn't be in there.

agarny replied:
Ok, reverted this in commit a270779c.

agarny committed:
Test utils: replaced the `timeit()` method with `timeNow()` and `elapsedTime()`.

- The `timeit()` method uses a method that is not portable while we now rely on C++11 to time things accurately.
- We don't have one method only anymore since this required passing a very specific type of function while we should be allowed to time anything we want.

agarny committed:
Test utils: deleted `owningModel()` and `owningComponent()`.

Indeed, those methods have been copied/pasted from our (main) utilities while we really ought to be using the (main) version of those methods.

agarny committed:
Annotator test: put it in the `utils` category.

agarny committed:
Test utils: replaced the `timeit()` method with `timeNow()` and `elapsedTime()`.

- The `timeit()` method uses a method that is not portable while we now rely on C++11 to time things accurately.
- We don't have one method only anymore since this required passing a very specific type of function while we should be allowed to time anything we want.

agarny committed:
Entity interface: removed a commented-out header.

agarny committed:
Enums interface: removed some commented-out code.

agarny committed:
Python bindings CMake file: removed some commented-out code.

agarny committed:
Units: simplified `setUnitId()`.

agarny committed:
Units: simplified `setUnitId()`.

agarny committed:
Utilities: some minor cleaning up.

hsorby commented:
I cannot fathom why these recent changes have been added to this pull request @agarny.  Clearly they are separate changes with nothing to do with the Annotator class.

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle committed:
fixed formatting in commonutils.h

kerimoyle commented on src/api/libcellml/module/libcellml:
```diff
@@ -29,6 +29,7 @@ limitations under the License.
 #include "libcellml/analysermodel.h"
 #include "libcellml/analyservariable.h"
+#include "libcellml/annotator.h"
 #include "libcellml/component.h"
```
Does this need `enums.h` too?

agarny replied:
`enums.h` is included by `types.h` which in turn is included by most headers in this module. So, no need to include `enums.h`.

agarny replied:
If we are to add `enums.h` then why don't we add `types.h`?

kerimoyle commented on src/bindings/interface/annotator.i:
```diff
@@ -0,0 +1,217 @@
+
+%feature("docstring") libcellml::Annotator::model
+"Return the model assigned to this Annotator or the model with the given id.";
+
```
Is this too ambiguous?  Should we have separate functions for these two things?

hsorby replied:
'Tis the beauty of function overloading.

kerimoyle replied:
Yeah .. but it sure ain't beautiful here :| The two functions are doing two totally different things!

agarny replied:
Any final decision on this?

hsorby replied:
I am happy as it currently is.

kerimoyle replied:
What happens if the current model in the annotator doesn't have the id specified?  Which function is it calling then?  I really don't like this.

agarny replied:
@hsorby and @kerimoyle: you guys need to come up with some kind of an agreement so that we can resolve this conversation. :)

nickerso replied:
> What happens if the current model in the annotator doesn't have the id specified? Which function is it calling then? I really don't like this.

Isn't this just pointing to the overloaded C++ code, so as with the C++ code if you ask for a model with a given id, its only going to return a non-null pointer if the id matches that of the current model. Right? The only difference here is that there is a fall back option to provide the current model for the annotator...

I don't see a need to separate this into two separate python methods. (assuming I'm understanding this issue correctly)

hsorby replied:
You have the same understanding that I have and I am happy with the way it works.

agarny replied:
I am happy with this too, so resolving.

kerimoyle commented on src/bindings/interface/annotator.i:
```diff
@@ -0,0 +1,217 @@
+
+%feature("docstring") libcellml::Annotator::unit
+"Return the UnitsPtr and index whose Unit is with given id.";
+
```
Is this still true?  Or do we have a single structure that's returned now?

agarny replied:
I believe this has been resolved.

kerimoyle commented on src/bindings/interface/annotator.i:
```diff
@@ -0,0 +1,217 @@
+
+%feature("docstring") libcellml::Annotator::connection
+"Return a pair of VariablePtrs that define a connection with given id.";
+
```
Ditto here - doesn't it now return a `VariablePair` item rather than two `VariablePtr`s?

agarny replied:
I believe this has been resolved.

kerimoyle commented on src/bindings/interface/annotator.i:
```diff
@@ -0,0 +1,217 @@
+"Clear all the id strings in the given or stored model.";
+
+%feature("docstring") libcellml::Annotator::isUnique
+"Return `true` if the given string is not duplicated in the stored model, or `false` otherwise.";
```
Can we reword this?  "Not duplicated" includes "does not exist".  Should rather be:
"Return `true` if the given string occurs exactly once as an id in the stored model, or `false` otherwise."

kerimoyle commented on src/bindings/interface/annotator.i:
```diff
@@ -0,0 +1,217 @@
+"Assign an automatically generated, unique id to the given component's encapsulation.";
+
+%feature("docstring") libcellml::Annotator::assignConnectionId
+"Assign an automatically generated, unique id to the connection between the given variables.";
```
"Assign an automatically generated, unique id to the connection between the parent components of the given variables.";

hsorby committed:
Improve documentation of Annotater class.

hsorby committed:
Merging in changes from 'prime/develop' fixing conflicts.

hsorby committed:
Merge branch 'annotator_use_any_not_variant' of github.com:kerimoyle/libcellml into kerimoyle/annotator_use_any_not_variant

agarny reviewed (commented):
> I cannot fathom why these recent changes have been added to this pull request @agarny. Clearly they are separate changes with nothing to do with the Annotator class.

Which ones are you talking about exactly? Those changes are the result of my reviewing this PR. So, all I have done is to refactor, clean up, etc. code that is in this PR.

agarny committed:
Merge branch 'annotator_use_any_not_variant' of https://github.com/kerimoyle/libcellml into pr/640

hsorby committed:
Modify tests to adjust for change in defaults in setting id for connections.

hsorby committed:
Merge branch 'annotator_use_any_not_variant' of github.com:kerimoyle/libcellml into kerimoyle/annotator_use_any_not_variant

hsorby committed:
Address review comments.

hsorby commented:
As far as I am concerned this work is ready for a full and proper review now.  (Assuming I haven't missed any comments) All comments have been addressed.

agarny commented:
> As far as I am concerned this work is ready for a full and proper review now. (Assuming I haven't missed any comments) All comments have been addressed.

Great, I will resume my review sometime today then. 👍

kerimoyle committed:
Merge branch 'develop' into annotator_use_any_not_variant

kerimoyle commented on src/bindings/interface/enums.i:
```diff
@@ -0,0 +1,26 @@
+    tmpD = {k:v for k,v in globals().items() if k.startswith(prefix + '_')}
+    tmpD = {k[len(prefix)+1:]:v for k,v in tmpD.items()}
+    globals()[prefix] = IntEnum(prefix,tmpD)
+redo('CellMLElement')
```
Needs renaming to `ItemType` to include XML.

hsorby replied:
This doesn't need to be done here.  It can come in with Issues.

kerimoyle commented on src/bindings/interface/types.i:
```diff
@@ -126,3 +149,15 @@ Only meant to be included, shouldn't be passed to cmake as a module!
+    %argument_fail(ecode, "$type", $symname, $argnum);
+  } else {
+    if (val < %static_cast($type::COMPONENT, int) || %static_cast($type::VARIABLE, int) < val) {
+      %argument_fail(ecode, "$type is not a valid value for the enumeration.", $symname, $argnum);
```
Needs to extend to use XML as last type

hsorby replied:
XML isn't required by the Annotator class so it should only come in when needed.

kerimoyle reviewed (commented):
I've only had a quick look, but my biggest concern is that we'll have to redo the `CellMLElement` enum immediately so that it doesn't entirely duplicate the `ItemType` one.  My comments from yesterday also seem to have been missed?

hsorby committed:
Update annotator.i

Changes to the method documentation.

nickerso commented on src/api/libcellml/annotator.h:
```diff
@@ -0,0 +1,1115 @@
+     * 
+     * @return the number of items with the given @p id.
+     */
+    size_t duplicateCount(const std::string &id);
```
perhaps not the best name for this method :) but can be fixed in future.

hsorby replied:
Pretty easy to fix now if you have one ready to go.

kerimoyle replied:
It has to end in `Count()`
It's returning the number of ids which are duplicated ... so ... `duplicateIdCount`?

agarny replied:
Why not simply `idCount()`?

kerimoyle replied:
... because that's not what it's doing!  It's returning the number of ids in the model which are duplicated.  Not the number of ids in the model.

agarny replied:
I understand that, but you are passing an id. So, if I was to have a call like `idCount("my_id")` then I would understand it as being the number of times `my_id` is used as an id. Alternatively, I think I would be fine with `duplicatedIdCount()` .

kerimoyle replied:
Maybe `itemCount` then? To me id count is counting the number of different ids.

nickerso replied:
I vote for `itemCount`

agarny replied:
> Maybe `itemCount` then? To me id count is counting the number of different ids.

Fair point, I would personally be happy with `itemCount()`.

agarny replied:
Are we ok with `itemCount()`? I am happy to modify it as part of my reviewing.

nickerso replied:
3 of us are happy with `itemCount` and no votes against so I think we're safe to go for it...

agarny replied:
Done in commit 47839635.

nickerso reviewed (dismissed):
I think this one is looking good and behaves as I think it should be. A couple of outstanding comments from @kerimoyle that should be resolved, but otherwise good to go.

kerimoyle commented on tests/annotator/annotator.cpp:
```diff
@@ -0,0 +1,2131 @@
+            case libcellml::CellMLElement::MAP_VARIABLES:
+                pair = std::any_cast<libcellml::VariablePair>(expectedItems[id][index].second);
+                testPair = std::any_cast<libcellml::VariablePair>(item.second);
+                // TODO not sure how best to test these.
```
remove?

agarny replied:
Yes. Done it as part of my reviewing of this PR.

agarny committed:
Types interface: some minor formatting.

agarny committed:
Annotator: make some tests easier to read...

... for me, at least.

agarny committed:
Annotator private implementation: use a const reference to pass some parameters.

agarny committed:
Annotator: removed unused variables.

agarny committed:
Annotator: make some tests easier to read...

... for me, at least.

agarny committed:
Annotator: prefer pre-increment to post-increment.

agarny committed:
Annotator: simplified try...catch() statements.

Thus ends up being consistent with the other try...catch() statements `src/annotator.cpp`.

agarny committed:
Annotator: removed unused variables.

agarny committed:
Annotator: various simplifications to the code.

agarny commented on src/api/libcellml/annotator.h:
```diff
@@ -0,0 +1,1115 @@
+    /**
+     * @brief Retrieves an item with the given id string.
+     *
+     * The item returned is a @c std::pair containing:
+     *  - a @ref CellMLElement enum, and
+     *  - a @c std::any item containing the item.
```
From PR #696 (see https://github.com/cellml/libcellml/pull/696#discussion_r499883292), it would make more sense to do swap the items. IOW, I would have:
```
     * The item returned is a @c std::pair containing:
     *  - a @c std::any item containing the item; and
     *  - a @ref CellMLElement enum.
```

agarny replied:
Does anyone have a view on this? Happy to create a separate issue for it, if needed.

agarny committed:
Annotator: some minor cleaning up in our API documentation.

agarny committed:
Annotator API: removed trailing spaces.

agarny commented on src/annotator.cpp:
```diff
@@ -0,0 +1,1678 @@
+        issue->setLevel(Issue::Level::ERROR);
+        return false;
+    }
+    mPimpl->mModel = model;
```
I have seen the comment at https://github.com/cellml/libcellml/pull/640/files#r482607731, but does it really apply here?

kerimoyle replied:
I don't understand what you mean?

kerimoyle replied:
Fixed in 3d9e774a

agarny committed:
Annotator: slight updates to the API documentation.

agarny committed:
Annotator: use consistent casing.

agarny commented on src/api/libcellml/types.h:
```diff
@@ -70,4 +77,49 @@ using UnitsPtr = std::shared_ptr<Units>; /**< Type definition for shared units p
+ *  - a connection between parent components; or
+ *  - an equivalence between two variables;
+ */
+using VariablePair = std::pair<VariablePtr, VariablePtr>; /**<  */
```
What is that trailing `/**<  */`?

hsorby replied:
Should have been deleted.

agarny replied:
It has gone anyway, so resolved.

agarny committed:
Annotator: use consistent casing.

agarny committed:
Annotator: slight update to the interface documentation.

agarny committed:
Make ClangFormat happy.

agarny committed:
Annotator: make some tests easier to read.

agarny committed:
Annotator: fixed a copy/paste gone wrong.

agarny committed:
Annotator: fixed a copy/paste gone wrong.

agarny committed:
Annotator: make sure that we get 100% coverage.

Indeed, the default return statement in those modified methods were never reached.

agarny committed:
Annotator Python test: make it PEP8 compliant.

agarny commented on src/variable.cpp:
```diff
@@ -509,20 +509,23 @@ void Variable::setEquivalenceMappingId(const VariablePtr &variable1, const Varia
 void Variable::setEquivalenceConnectionId(const VariablePtr &variable1, const VariablePtr &variable2, const std::string &connectionId)
```
In other places, we use a `VariablePair` when wanting to do something for two variables. Why not here?

kerimoyle replied:
A discussion for another time ... I've made an issue for it at https://github.com/cellml/libcellml/issues/745

agarny commented on tests/annotator/annotator.cpp:
```diff
@@ -0,0 +1,2131 @@
+    auto v4 = libcellml::Variable::create("v4");
+    c1->addVariable(v3);
+    c2->addVariable(v4);
+    auto v1v2 = std::make_pair(c1->variable("v1"), c2->variable("v2"));
+    auto v3v4 = std::make_pair(c1->variable("v3"), c2->variable("v4"));
```
Exactly the kind of thing I wouldn't expect a libCellML user to have to do (i.e. use something like `std::make_pair()`).

agarny replied:
Will be addressed in issue #748.

agarny commented on tests/annotator/annotator.cpp:
```diff
@@ -0,0 +1,2131 @@
+    annotator->setModel(model);
+
+    // Unit items are returned as name string and pair of UnitsPtr parent and index:
+    EXPECT_EQ(libcellml::CellMLElement::UNIT, annotator->item("unit1_id").first);
+    EXPECT_EQ(units, std::any_cast<libcellml::UnitItem>(annotator->item("unit1_id").second).first);
+    EXPECT_EQ(size_t(0), std::any_cast<libcellml::UnitItem>(annotator->item("unit1_id").second).second);
```
Concrete example of what I mentioned about the use of `std::pair`. Here, we have no idea what `x.first` and `x.second` are about. We can tell from the context, but it's not ideal.

agarny replied:
Created issue #748 for it.

agarny commented on tests/annotator/annotator.cpp:
```diff
@@ -0,0 +1,2131 @@
+
+    annotator->setModel(model);
+
+    // Unit items are returned as name string and pair of UnitsPtr parent and index:
```
What do you mean by "name string" (same below)? From what I can tell, it's a `class enum`, no?

kerimoyle replied:
It is now :)

kerimoyle replied:
b386607e

agarny commented on tests/annotator/annotator.cpp:
```diff
@@ -0,0 +1,2131 @@
+    EXPECT_EQ(importSource, std::any_cast<libcellml::ImportSourcePtr>(annotator->item("import_id").second));
+}
+
+TEST(Annotator, getNonEntityFromId)
```
What is meant by non-entity from id?

hsorby replied:
Something that isn't a libCellML entity: VariablePair.

agarny replied:
Then, I woudl expect something like `getVariablePairFromId`. At least, I don't have to guess/ask/etc. what `NonEntity` is...

agarny replied:
Done in commit 2df300f4.

agarny commented on tests/annotator/annotator.cpp:
```diff
@@ -0,0 +1,2131 @@
+TEST(Annotator, duplicateIdBehaviour)
+{
+    std::vector<std::string> expectedErrors = {
+        "The id 'duplicateId' occurs 3 times in the model so a unique item cannot be located."};
```
I feel like the error message should read "... model, so..." (i.e. missing comma?).

kerimoyle replied:
I prefer it without.

agarny commented on tests/annotator/annotator.cpp:
```diff
@@ -0,0 +1,2131 @@
+    EXPECT_EQ("b4da62", annotator->assignId(itemConnection2));
+    EXPECT_EQ("b4da63", annotator->assignId(itemMapVariables2));
+
+    EXPECT_TRUE(annotator->hasModel());
```
Why test `annotator->hasModel()` so many since it's always `true`?

hsorby replied:
That is a search/replace special `hasModel` was originally something else and I reappropriated it.

agarny commented on tests/annotator/annotator.cpp:
```diff
@@ -0,0 +1,2131 @@
+
+TEST(Annotator, assignIdInvalidType)
+{
+    const std::string e = "The given type 'reset' is invalid for this method.";
```
This error message doesn't help me. I would expect to know why it is invalid.

kerimoyle replied:
b386607e

agarny committed:
Annotator tests: cleaned up the code and added a few more bits of tests here and there.

agarny committed:
Annotator: make some tests easier to read.

agarny reviewed (commented):
Looks ok overall, but still have a few more questions to which it would be nice to have an answer before approving. :)

kerimoyle committed:
fixed retrieval so that missing ids get the right error returned

kerimoyle committed:
Reinstate ability to pass an item tuple into the Annotator::assignId function

kerimoyle committed:
tweaked bindings for coverage

kerimoyle committed:
overload in python binding for assignId

kerimoyle committed:
uncommented tests, tidied up.

kerimoyle committed:
included issues in the annotator bindings

kerimoyle committed:
coverage, renamed tests

kerimoyle committed:
improved error messages

kerimoyle committed:
fix call to mModel instead of setModel

hsorby committed:
Remove repeated use of annotator->hasModel() that doesn't do the job it once did.

nickerso commented on src/annotator.cpp:
```diff
@@ -450,30 +446,33 @@ void Annotator::AnnotatorImpl::addIssueNoModel() const
 {
     AnyItem retrieved = std::make_pair(CellMLElement::UNDEFINED, nullptr);
-    if (isUnique(id)) {
+    auto num = duplicateCount(id);
```
hasn't `duplicateCount` been renamed?

agarny replied:
Nope, see https://github.com/cellml/libcellml/pull/640/files#r499497547. Some of use are fine with renaming it `itemCount`. So, am happy to make that change. Agreed?

agarny replied:
Done in commit 47839635.

nickerso reviewed (dismissed):
Thought we were moving away from `duplicateCount` but that can be resolved in future if its still an issue.

agarny committed:
Annotator: renamed the `getNonEntityFromId` test to `getVariablePairFromId`.

agarny committed:
CMake: reverted some in `cmake/environmentchecks.cmake`.

For some reasons, this was added, but has nothing to do with the issue at hand.

agarny committed:
Annotator: renamed `duplicateCount()` with `itemCount()`.

agarny committed:
Annotator interface: some minor cleaning up.

agarny committed:
Python bindings: make PEP8 happy.

agarny reviewed (dismissed):
There are still a few unresolved conversations, but none that I am concerned with. So, personally, I am happy with the PR as-is.

agarny commented:
> There are still a few unresolved conversations, but none that I am concerned with. So, personally, I am happy with the PR as-is.

Oops, I have somehow broken this PR. Argh! :)

agarny committed:
Revert "Python bindings: make PEP8 happy."

This reverts commit cf5ab33ac5069b8f06f76e31d4b5d37672848af5.

agarny commented:
@hsorby, @kerimoyle and @nickerso: any idea why Ihad to revert commit cf5ab33 to get the Python tests to run fine again? I really don't like the idea of having to import `CellMLElement` before anything else for things to work. This is really bad, not to mention that I like, in PyCharm, to reformat my code. So, I will regularly end up with something like in commit cf5ab33, which will then break things as we have seen.

IOW, I would like to have something like commit cf5ab33 back in before we merge this PR.

agarny commented:
> IOW, I would like to have something like commit [cf5ab33](https://github.com/cellml/libcellml/commit/cf5ab33ac5069b8f06f76e31d4b5d37672848af5) back in before we merge this PR.

Ok, I had overlooked the fact that it's in our `__init__.py`. So, it's not like an end-user will have to remember to import `CellMLElement` first. So, ok, although I don't like the fact that we import `CellMLElement` first, it doesn't affect the end-user, so I guess I am "happy" with the PR as-is.
